Create 2020-11-19-error-message-design.md (#200)
temporalio/documentation: blog/2020-11-19-error-message-design.md

---
tags:
- developer-experience
- errors
posted_on_: 2020-11-19T23:06:09Z
id: error-message-design

> title: 'Write Errors That Don't Make Me Think'

Author: ```suggestion
title: Write Errors That Don't Make Me Think
```

author: Swyx (Shawn Wang)
author_title: Product Manager
author_image_url: https://avatars2.githubusercontent.com/u/6764957?s=460&u=97ad815028595b73b06ee4b0510e66bbe391228d&v=4
release_version: V1.3

---

> 

Reviewer: You're going to need a truncate tag or the entire blog content will show up in the feed.

```suggestion
<!--truncate-->
```

There is nothing more frustrating than a cryptic error message. 

When something goes wrong, chances are you already knew it - all a cryptic error message does is confirm your suspicion *and* tell you that you're gonna need to read through a bunch of code to figure out your error.

> As framework and library designers, we can and should do better for our users with some careful design principles.

Reviewer: ```suggestion
As framework and library designers, we can and should do better for our users by carefully following a few design principles.
```

## Why We Should Care

> Who among us hasn't thrown up their hands at an unhelpful `Error: NullPointerException` or `undefined is not a function`? Of course, this is a [straw man](https://en.wikipedia.org/wiki/Straw_man) - unanticipated errors always leak internals. We should do our best to handle these errors as much as possible! 

Reviewer: The straw man analogy is a bit rough

The frustrating part comes from errors that were *anticipated* but obviously not given any thought:

> ![Alt Text](https://dev-to-uploads.s3.amazonaws.com/i/qiivj3cdvkbo07fwrxq6.png)

Author: ```suggestion
![Example of unhelpful error message from Twitter screenshot](https://dev-to-uploads.s3.amazonaws.com/i/qiivj3cdvkbo07fwrxq6.png)
```

This is particularly aggravating because it compounds an already unhappy situation with an even worse debugging and bug reporting experience.

Thoughtful error message design is one of the less glamorous, but most important, parts of [Developer Exception Engineering](https://www.swyx.io/developer-exception/). Most framework developers instinctively write error messages that make sense to *them*, but a few minutes of extra effort can save DAYS of debugging for *everyone else* (including maintainers!). 

We must acknowledge two things:

> - there is immense information asymmetry between framework developer and framework user (developers know the codebase far better than users)

Reviewer: ```suggestion
- there is immense information asymmetry between the framework developer and the framework user (developers know the codebase far better than users)
```

- error messages are part of your developer experience too

> You might be familiar with the old adage "If you fail to plan, you plan to fail." Here, we have a slight twist: **If you fail to plan to fail, you are planning to fail your users when they need you most.** Touting great developer experience with lousy error message design is like selling a beautiful car with seatbelts that don't work when there's a crash.

Reviewer: I was doubtful but the seatbelt analogy is really good.

## Error Message Style Guide

At Temporal, our current style guide reflects our engineering roots in [Uber](https://github.com/uber-go/guide/blob/master/style.md#error-types) and [Golang](https://blog.golang.org/go1.13-errors) with some tactical recommendations:

> - Use `errors.New` for simple messages and `fmt.Errorf` for formatted messages. Make sure `errors` is standard go package.

Reviewer: ```suggestion
- Use `errors.New` for simple messages and `fmt.Errorf` for formatted messages. Make sure you're using the standard `errors` go package.
```

- Error messages start with lowercase and no punctuation at the end: `unable to open file` instead of `Unable to open file..`
- Error messages should be actionable, e.g. `no such directory` instead of `loading service config`.

> - Use `unable to...` instead of `failed to...`, `can't...`, or `could not...`.

Reviewer: No issues but curious rationale. Because its neutral language instead of negative?

Author: honestly idk i just copied it from our own style guide

> - Messages should reflect logic action that we just tried to make (in a call on previous line which returned an error), not a function name: `load service config` instead of `LoadServiceConfig`. This message should add actionable meaningful context to the existing error message.

Reviewer: I feel this is the weakest since the wording doesn't make message as clear as it could be.

Author: ```suggestion
- Messages should reflect the logical action, not merely a function name: `load service config` instead of `LoadServiceConfig`. This message should add actionable, meaningful context to the existing error message.
```

> That said, as we look to evolve our philosophy of developer experience, we are looking toward higher level design principles that help guide successful outcomes.

Reviewer: ```suggestion
That said, as we evolve our developer experience philosophy, we are looking toward higher level design principles that help guide successful outcomes.
```

## Error Message Design Principles 

> We are still in the early stages of pinning down our beliefs in what makes for great error messages, but here are some candidate principles running through my mind:

Reviewer: High level feedback:

* Might be nice to include some example errors which showcase these tenants
* I would also add "no jargon". Can't tell you how many times 
* No misleading errors! They are worse than no errors or errors which are useless. There is an error called "unresolved external symbol" in C/C++ which is seriously the most frustrating error for beginners ever. It never points you to where the actual problem is. 
* Errors should be reserved for errors unless user explicitly configures otherwise. Many users don't care about warnings for better and worse
* Huge stack traces suck especially when I only care about the line at the bottom
* I sort of think that a good error experience should be like a great testing framework. This goes towards your errors should display expected vs receive. This is exactly like a good testing framework which shows the literal diff between expected and received. I would love that in an error framework!

- **Errors should be written for USERS, not maintainers**: Your error should make sense from a user point of view. For example, you should cite user-supplied configs and values familiar to users, not the internal variable and function names familiar to maintainers. 

> - **Errors should suggest WHY they happened**, if not HOW to fix them: As a maintainer, you probably have a high amount of context as to the possible causes of an error - take the time to write down some hints for your users. If there's a high probability you know what's wrong, the error message itself can suggest how to fix it. But even without that certainty, think about what the user is likely to need to log out to debug the error, and offer it in the error message itself. Where relevant, **Errors should display what was expected vs what was received**.

Reviewer: ```suggestion
- **Errors should suggest WHY they happened**, or even better - HOW to fix them: As a maintainer, you probably have a high amount of context as to the possible causes of an error - take the time to write down some hints for users. If there's a high probability you know what's wrong, the error message itself can suggest how to fix it. But even without that certainty, think about what the user is likely to need to log out to debug the error, and offer it in the error message itself. Where relevant, **Errors should display what was expected vs what was received**.
```
" think about what the user is likely to need to log out to debug the error, and offer it in the error message itself."

English is wrong I think
